[SR-38][Sema] Stop emitting duplicate only-as-generic-constraint errors.

gregomni opened the pull request:
There are two separate issues here. First, top-level variable declarations were being checked once by the Decl checker, and then again by the Stmt checker. The Decl checker now skips global VarDecls. This fixes SR-38.

Second, the Stmt checker is called by an AST visitor itself, which already calls it once per statement. Using the UnsupportedProtocolVisitor here meant that each interior sub statement would get visited multiple times. Added a setRecurseIntoSubstatements() on the visitor, and set it to false for the Stmt checker. This fixes the pre-existing testHasAssoc() which used to produce 2 errors.

slavapestov commented on lib/Sema/TypeCheckType.cpp:
```diff
@@ -3453,6 +3472,11 @@ void TypeChecker::checkUnsupportedProtocolType(Decl *decl) {
       decl->getDeclContext()->isModuleScopeContext())
     return;
 
+  // Global variable declarations are okay, because they'll be caught in the stmt checker.
```
Is this true even in -parse-as-library mode?

gregomni replied:
@slavapestov Yes, just verified that. In -parse-as-library mode there isn't a top-level statement, but there is a top-level pattern decl instead, and the checking on that pattern decl catches the unsupported protocol type in the same source code location. I've updated the comment here to be more accurate.

slavapestov commented on lib/Sema/TypeCheckType.cpp:
```diff
@@ -3453,6 +3472,13 @@ void TypeChecker::checkUnsupportedProtocolType(Decl *decl) {
       decl->getDeclContext()->isModuleScopeContext())
     return;
 
+  // Global variable declarations are okay to skip, because at the top level of script/repl
```
So it sounds like you're actually skipping decls nested inside statements or other decls here, because those decls will be visited later?

gregomni replied:
Right, indirectly. The existing DeclChecker only calls this code (the decl version of checkUnsupportedProtocolType()) if it is doing a second pass or is in a protocol or extension context, and it only does a second pass over things at module context. So it isn't getting called anywhere where there could be statements, except for TopLevelCodeDecls when those are allowed.

But it doesn't skip nested decls. So talking through this has also led me to another couple cases of doubling (trebling, quadrupling...):
`protocol P {
    typealias T
}
protocol Q {
    var s: P { get set }
}` ... which will currently complain about protocol P four times, once for the PatternDecl, once for the VarDecl, and then once each for the anonymous get and set FuncDecls.

And: `extension Q {
    func bar(x: P) {}
}` ... which will complain twice, once for the top-level ExtensionDecl and once for the FuncDecl inside.

So clearly there is more to be done here. Thanks for the leading questions. :)

slavapestov replied:
No worries. Also don't feel like you have to get everything right 100%, because clearly it was broken before, and even if its not perfect we could check something in that makes it less broken :-)

The only case I'd be concerned about is if your changes accidentally made it so that we were not emitting a diagnostic in a case where we should be -- passing down an invalid existential to SILGen and IRGen will likely cause crashes.

gregomni replied:
@slavapestov Ok! Committed fixes for all the bits identified above.

gregomni commented:
Ended up backing off one of the changes because it caused a couple testing failures, and added a FIXME comment instead to replace it. Still, this gets almost all the way there. Committed, squashed, all tests pass.

lattner commented:
The TypeCheckType.cpp/TypeCheckDecl.cpp changes look unrelated to this work?

gregomni commented:
That change is to only call TC.checkUnsupportedProtocolType() once for decls inside a protocol context. Previously it would call for both IsFirstPass and IsSecondPass if two were needed. The new if statement only does it on the first pass or only pass.

lattner commented:
Ok, but why is that part of this patch?  Is that to avoid redundant diagnostics?  If so, can you split that out into something that stands by itself?

gregomni commented:
Why: uh... hubris? :-) Rolled back that part and resquashed.

lattner commented:
Thanks!  This patch looks fine to me, but have you considered an approach where you mark the type repr as invalid when this happens?  Something like this:

  void visitIdentTypeRepr(IdentTypeRepr *T) {
    auto comp = T->getComponentRange().back();
    if (auto proto = dyn_cast_or_null<ProtocolDecl>(comp->getBoundDecl())) {
      if (!proto->existentialTypeSupported(&TC) && !T->isInvalid()) {
        TC.diagnose(comp->getIdLoc(), diag::unsupported_existential_type,
                    proto->getName());
        T->setInvalid();
        Diagnosed.insert(proto);
      }

      return;
    }

lattner commented:
However, I agree with you that making "checkUnsupportedProtocolType(Stmt *stmt)" avoid walking substatements makes sense, just as an efficiency win.

gregomni commented:
I haven't investigated invalidating the type repr, but yeah, that seems very promising to help make sure each one is only diagnosed once, even if a visitor happens to go over it again. It would certainly be more resilient to future change in the rest of the type checker than the carve out skip-these checks that are there now.

But I think, if I remember correctly, that the remaining duplicates are actually from the other diagnose call site (line 3505), where we check types of ValueDecls. That is catching - I think - protocol composition types and type aliases to them. Probably the way to go is to invalidate type repr's as you suggest, but the logic in UnsupportedProtocolVisitor needs to be beefed up to catch more kinds of type reprs, and then the value decl type check can go away entirely. (Which would be an additional win, because where those are hit right now the diagnosis points to the value decl's identifier instead of to the type repr.)

gregomni commented:
Okay! New commit (and squashed, and tests pass) using invalidation of the type repr to stop duplications instead of fragile checks on decl type. Also moved the code that was checking the type of ValueDecls into checking the underlying type of TypeAliasDecls, which turned out to be the actual specific place that this checking was required.

Figuring that out revealed a hole in the previous checks. If you used a typealias inside statements, whose underlying type was a protocol composition that contained an unsupported protocol, then the current master would fail to diagnose it. Added this as a test at the end of test/type/protocol_types.swift.

slavapestov commented:
Excellent work, thanks!
